org-brain.el

```diff
@@ -93,39 +165,54 @@ If RELATIVE is t, then return relative paths and remove org extension."
 
 (defun org-brain-parents (entry)
   "Get list of org-brain entries which links to ENTRY."
-  (-non-nil
-   (mapcar
-    (lambda (brainfile)
-      (let ((brainfile-entry (org-brain-path-entry-name brainfile)))
-        (unless (string-equal brainfile-entry entry)
-          (org-element-map
-              (with-temp-buffer
-                (insert-file-contents brainfile)
-                (org-element-parse-buffer))
-              'link
-            (lambda (link)
-              (when (and (string-equal (org-element-property :type link) "brain")
-                         (string-equal (car (split-string (org-element-property :path link) "::"))
-                                       entry))
-                brainfile-entry))
-            nil t))))
-    (org-brain-files))))
+  (if (and org-brain-parents-cache
+           (assoc entry org-brain-parents-cache))
+      (cdr (assoc entry org-brain-parents-cache))
+    (progn
+      (org-brain-log (format  "Updating org-brain-parents-cache for %s..." entry))
+      (let ((parents (remove nil
+                      (mapcar
+                       (lambda (brainfile)
+                         (let ((brainfile-entry (org-brain-path-entry-name brainfile)))
+                           (unless (string-equal brainfile-entry entry)
+                             (org-element-map
+                                 (with-temp-buffer
+                                   (insert-file-contents brainfile)
+                                   (org-element-parse-buffer))
+                                 'link
+                               (lambda (link)
+                                 (when (and (string-equal (org-element-property :type link) "brain")
+                                            (string-equal (car (split-string (org-element-property :path link) "::"))
+                                                          entry))
+                                   brainfile-entry))
+                               nil t))))
+                       (org-brain-files)))))
+        (push (cons entry . (parents)) org-brain-parents-cache)
+        (cdr (assoc entry org-brain-parents-cache))))))
 
 (defun org-brain-children (entry &optional exclude)
   "Get list of org-brain entries linked to from ENTRY.
 You can choose to EXCLUDE an entry from the list."
-  (delete
-   exclude
-   (delete-dups
-    (org-element-map
-        (with-temp-buffer
-          (ignore-errors (insert-file-contents (org-brain-entry-path entry)))
-          (org-element-parse-buffer))
-        'link
-      (lambda (link)
-        (when (string-equal (org-element-property :type link) "brain")
-          (let ((link-entry (car (split-string (org-element-property :path link) "::"))))
-            (unless (string-equal link-entry entry) link-entry))))))))
+  ;; TODO Handle exclude
```

Comment:
Not sure if we need exclude. I think I required it for a reason at first, but it might be so that I ended up not using it in the codebase :)